### PR TITLE
Add retire warning to legacy code

### PR DESCRIPTION
**Pull request by zhangguanheng66**

### CODE REVIEW

File: torchtext/data/batch.py [hunk 1]
```diff
@@ -19,6 +20,7 @@ class Batch(object):
 
     def __init__(self, data=None, dataset=None, device=None):
         """Create a Batch from a list of examples."""
+        warnings.warn('Batch class will retire in 0.8.0 release', RuntimeWarning)
```
Comment (mthrok) [inline torchtext/data/batch.py:23]: According to [this comment](https://github.com/pytorch/audio/pull/523#discussion_r405673957), PyTorch's convention for deprecation is `UserWarning`.

File: torchtext/data/example.py [hunk 2]
```diff
@@ -9,6 +10,7 @@ class Example(object):
     """
     @classmethod
     def fromJSON(cls, data, fields):
+        warnings.warn('Example class will retire in 0.8.0 release', RuntimeWarning)
```
Comment (vincentqb) [inline torchtext/data/example.py:13]: For all of these warnings, the user could ask themselves many questions. Is there any context that should be given? What is the user to do instead? Should the user file an issue if they can't do what they want?

File: torchtext/data/batch.py (hunk 1 above)
Comment (zhangguanheng66) [reply torchtext/data/batch.py:23]: Thanks

File: torchtext/data/example.py (hunk 2 above)
Comment (zhangguanheng66) [reply torchtext/data/example.py:13]: We have an issue to explain this change. I might add this to the warning message. And in 0.7.0 release note, I will introduce this again.

Comment (vincentqb) [reply torchtext/data/example.py:13]: Would adding the link in the message be a good idea? btw can you share the link here in the comment? :)

Comment (zhangguanheng66) [reply torchtext/data/example.py:13]: https://github.com/pytorch/text/issues/664

Comment (vincentqb) [reply torchtext/data/example.py:13]: Based on offline discussion: Please use a standard message for all. The message needs to say: (1) users are recommended to move to the new interface by doing abc, but (2) they can keep using the same functionality without the warning by doing xyz.

File: torchtext/data/batch.py [hunk 3]
```diff
@@ -19,6 +20,7 @@ class Batch(object):
 
     def __init__(self, data=None, dataset=None, device=None):
         """Create a Batch from a list of examples."""
+        warnings.warn('Batch class will retire in 0.8.0 release and stay in torchtext.legacy. See 0.7.0 release note for the replacement.', UserWarning)
```
Comment (vincentqb) [inline torchtext/data/batch.py:23]: Can you add the link to the release notes?

Comment (zhangguanheng66) [reply torchtext/data/batch.py:23]: For now, we don't have a link. And I'm thinking if it's proper to add a link in warning message.

Comment (vincentqb) [reply torchtext/data/batch.py:23]: That's a good question, but I'm also not a fan of mentioning the release notes alone :)

File: torchtext/data/example.py [hunk 4]
```diff
@@ -9,6 +10,7 @@ class Example(object):
     """
     @classmethod
     def fromJSON(cls, data, fields):
+        warnings.warn('Example class will retire in 0.8.0 release and stay in torchtext.legacy. See 0.7.0 release note for the replacement.', UserWarning)
```
Comment (cpuhrsch) [inline torchtext/data/example.py:13]: I'd change the phrasing of this a bit.

"Example class will retire in 0.8.0 release and stay in torchtext.legacy. See 0.7.0 release note for the replacement"

to 

"Example class will be retired in the 0.8.0 release and moved to torchtext.legacy. Please see 0.7.0 release notes for further information."

Comment (cpuhrsch) [reply torchtext/data/example.py:13]: Did we settle on UserWarning? cc @gchanan

Comment (zhangguanheng66) [reply torchtext/data/example.py:13]: With some offline discussions, `UserWarning` is proper, though pytorch started to use `FutureWarning`.

File: torchtext/data/example.py [hunk 5]
```diff
@@ -9,6 +10,7 @@ class Example(object):
     """
     @classmethod
     def fromJSON(cls, data, fields):
+        warnings.warn('Example class will be retired in the 0.8.0 release and moved to torchtext.legacy. Please see 0.7.0 release note for further information.', UserWarning)
```
Comment (cpuhrsch) [inline torchtext/data/example.py:13]: nit: It's "notes" not "note".

**Pull request merged**